File: internal/filter/filter.go
package filter

type CompOp string

const (
	CompOpEq   CompOp = "="
	CompOpNeq  CompOp = "!="
	CompOpLike CompOp = "~="
	CompOpLt   CompOp = "<"
	CompOpLe   CompOp = "<="
	CompOpGt   CompOp = ">"
	CompOpGe   CompOp = ">="
)

func (op CompOp) Precedence() int {
	return 3
}

type BoolOp string

const (
	BoolOpAnd BoolOp = "and"
	BoolOpOr  BoolOp = "or"
)

func (op BoolOp) Precedence() int {
	switch op {
	case BoolOpAnd:
		return 2
	case BoolOpOr:
		return 1
	default:
		return 0
	}
}

type Visitor interface {
	VisitCompFilter(*CompFilter) (Visitor, error)
	VisitBoolFilter(*BoolFilter) (Visitor, error)
}

type Filter interface {
	Accept(Visitor) error
	Precedence() int
}

type CompFilter struct {
	Field string
	Op    CompOp
	Value Value
}

func (c *CompFilter) Accept(visitor Visitor) error {
	if _, err := visitor.VisitCompFilter(c); err != nil {
		return err
	}
	return nil
}

func (c *CompFilter) Precedence() int {
	return c.Op.Precedence()
}

type BoolFilter struct {
	LHS Filter
	Op  BoolOp
	RHS Filter
}

func (b *BoolFilter) Accept(visitor Visitor) error {
	v, err := visitor.VisitBoolFilter(b)
	if err != nil {
		return err
	}
	if v == nil {
		return nil
	}
	if b.RHS != nil {
		if err := b.RHS.Accept(v); err != nil {
			return err
		}
	}
	if b.LHS != nil {
		if err := b.LHS.Accept(v); err != nil {
			return err
		}
	}
	return nil
}

func (b *BoolFilter) Precedence() int {
	return b.Op.Precedence()
}

func And(lhs, rhs Filter, filters ...Filter) Filter {
	filter := &BoolFilter{lhs, BoolOpAnd, rhs}
	for _, f := range filters {
		filter = &BoolFilter{filter, BoolOpAnd, f}
	}
	return filter
}

func Or(lhs, rhs Filter, filters ...Filter) Filter {
	filter := &BoolFilter{lhs, BoolOpOr, rhs}
	for _, f := range filters {
		filter = &BoolFilter{filter, BoolOpOr, f}
	}
	return filter
}

func Eq(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpEq, Value: val}
}

func Neq(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpNeq, Value: val}
}

func Gt(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpGt, Value: val}
}

func Ge(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpGe, Value: val}
}

func Lt(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpLt, Value: val}
}

func Le(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpLe, Value: val}
}

func Like(field string, val Value) Filter {
	return &CompFilter{Field: field, Op: CompOpLike, Value: val}
}

File: gui/src/app/lib/filter-dsl/filter.ts
import { Tag, Tok } from './scanner';
import { Value } from './value';

export enum CompOp {
  Eq = '=',
  Neq = '!=',
  Like = '~=',
  Lt = '<',
  Le = '<=',
  Gt = '>',
  Ge = '>=',
}


function compOpPrecedence(op: CompOp): number {
  return 3;
}

export enum BoolOp {
  And = 'and',
  Or = 'or',
}

function boolOpPrecedence(op: BoolOp): number {
  switch (op) {
    case BoolOp.And:
      return 2;
    case BoolOp.Or:
      return 1;
  }
  return 0;
}

export function isBoolOp(token: Tok) {
  return token.tag === Tag.And || token.tag === Tag.Or;
}

export interface Visitor {
  visitCompFilter(f: CompFilter): Visitor

  visitBoolFilter(f: BoolFilter): Visitor
}

export interface Filter {
  accept(v: Visitor): void

  precedence(): number
}

export class CompFilter implements Filter {

  constructor(public field: string, public op: CompOp, public value: Value) {
  }

  accept(v: Visitor): void {
    v.visitCompFilter(this);
  }

  precedence(): number {
    return compOpPrecedence(this.op);
  }
}

export class BoolFilter implements Filter {

  constructor(public lhs: Filter, public op: BoolOp, public rhs: Filter) {
  }

  accept(v: Visitor): void {
    v.visitBoolFilter(this);
  }

  precedence(): number {
    return boolOpPrecedence(this.op);
  }
}

export function NewCompFilter(field: string, op: CompOp, value: Value): CompFilter {
  return new CompFilter(field, op, value);
}

export function And(lhs: Filter, rhs: Filter, ...filters: Filter[]): Filter {
  let filter = new BoolFilter(lhs, BoolOp.And, rhs);
  filters.forEach((f: Filter) => {
    filter = new BoolFilter(filter, BoolOp.And, f);
  });
  return filter;
}

export function Or(lhs: Filter, rhs: Filter, ...filters: Filter[]): Filter {
  let filter = new BoolFilter(lhs, BoolOp.Or, rhs);
  filters.forEach((f: Filter) => {
    filter = new BoolFilter(filter, BoolOp.Or, f);
  });
  return filter;
}


export function Eq(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Eq, value);
}

export function Neq(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Neq, value);
}

export function Gt(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Gt, value);
}

export function Ge(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Ge, value);
}

export function Lt(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Lt, value);
}

export function Le(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Le, value);
}

export function Like(field: string, value: Value): Filter {
  return new CompFilter(field, CompOp.Like, value);
}

